### PR TITLE
Add a commandline server test runner

### PR DESCRIPTION
**Pull request by ben-clayton**
This is similar to the cmdline tool, but acts a simple HTTP RPC test runner.

This is significantly faster than using the cmdline runner to run each test in isolation, as it remove the overhead of scanning for tests, and constructing a new device each test.

### CODE REVIEW

File: src/common/runtime/server.ts [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+      }
+    }
+  );
+
+  server.listen(0, () => {
+    const address = server.address() as AddressInfo;
+    console.log(`Server listening at [[${address.port}]]`);
+  });
+})().catch(ex => {
+  console.log(ex.stack ?? ex.toString());
```
Comment (austinEng) [inline src/common/runtime/server.ts:115]: console.error?
```suggestion
  console.error(ex.stack ?? ex.toString());
```

File: src/common/runtime/server.ts [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+    return res;
+  }
+
+  const server = http.createServer(
+    async (request: http.IncomingMessage, response: http.ServerResponse) => {
+      if (request.url === undefined) {
+        return;
+      }
+      if (request.url.startsWith('/run?')) {
+        const name = request.url.substr(5);
```
Comment (austinEng) [inline src/common/runtime/server.ts:88]: maybe:
```suggestion
      const runPrefix = '/run?';
      if (request.url.startsWith(runPrefix)) {
        const name = request.url.substr(runPrefix.length);
```

File: src/common/runtime/server.ts [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+for (let i = 0; i < sys.args.length; ++i) {
+  const a = sys.args[i];
+  if (a.startsWith('-')) {
+    if (a === '--verbose') {
+      debug = true;
+    } else if (a === '--gpu-provider') {
+      const modulePath = sys.args[++i];
+      setGPUProvider(() => require(modulePath).gpu);
+    }
+  }
```
Comment (austinEng) [inline src/common/runtime/server.ts:53]: check for `--help` and print the usage info?

File: src/common/runtime/server.ts [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+}
+
+(async () => {
+  Logger.globalDebugMode = debug;
+  const log = new Logger();
+  const testcases = await allWebGPUTestcases();
```
Comment (austinEng) [inline src/common/runtime/server.ts:59]: nit: if we just have

`const testcasesPormise = allWebGPUTestcases();`

and later on inside the request handler do:

`const testcase = (await testcasesPromise).get(name);`

then the server will get instantiated and start listening right away, and block later on if testcases aren't done loading yet.
I find it a minor annoyance when I start a tool, hit the web server, and then realize it hasn't started listening yet.

File: src/common/runtime/server.ts [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+    const name = testcase.query.toString();
+    const [rec, res] = log.record(name);
+    await testcase.run(rec, expectations);
+    return res;
+  }
+
+  const server = http.createServer(
+    async (request: http.IncomingMessage, response: http.ServerResponse) => {
+      if (request.url === undefined) {
+        return;
```
Comment (austinEng) [inline src/common/runtime/server.ts:85]: when does this happen? should we return an error status code?

File: src/common/runtime/server.ts [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+          }
+        } else {
+          message = `test case '${name}' not found`;
+        }
+        const res: RunResult = { status, message };
+        response.end(JSON.stringify(res));
+      } else if (request.url.startsWith('/terminate')) {
+        server.close();
+        sys.exit(1);
+      }
```
Comment (austinEng) [inline src/common/runtime/server.ts:106]: error response if it doesn't match /run or /terminate? I think these requests will hang around if they're never answered

File: src/common/runtime/server.ts (hunk 5 above)
Comment (ben-clayton) [reply src/common/runtime/server.ts:88]: Done

File: src/common/runtime/server.ts (hunk 3 above)
Comment (ben-clayton) [reply src/common/runtime/server.ts:55]: Done

File: src/common/runtime/server.ts (hunk 2 above)
Comment (ben-clayton) [reply src/common/runtime/server.ts]: Done

**Pull request merged**